### DISCUSSION TITLE
Force Rendering of PDF Documents

### DISCUSSION THREAD

**Discussion by dbuechel** [cefsharp/CefSharp#4584] [Q&A] [answered]
We have a use case for which we always need to render PDFs inside the browser, even if the response has e.g. the content disposition set to "attachement" (and not "inline") or if the link to the PDF specifies the download attribute.

I already managed to detect a PDF document download, cancel it and spawn a new browser control with the URL of the PDF, but now it is unclear to me how I can force the rendering of a PDF document with the internal PDF viewer (PDFium)? Is there a special URL format or scheme I need to use (e.g. `pdf-extension://url-to-the-pdf-document`)?

**Comment by amaitland** [comment]
It's not clear what you've actually tried or the current behaviour you are seeing.

> but now it is unclear to me how I can force the rendering of a PDF document with the internal PDF viewer (PDFium)? Is there a special URL format or scheme I need to use (e.g. `pdf-extension://url-to-the-pdf-document`)?

What do you mean exactly? Just load the `PDF` in the browser as per normal.

> Is there a special URL format or scheme I need to use (e.g. `pdf-extension://url-to-the-pdf-document`)?

There is a scheme from memory, I've never tried to interact with it directly.

**Reply by dbuechel** [reply]
We have a custom download handler (see https://github.com/SafeExamBrowser/seb-win-refactoring/blob/master/SafeExamBrowser.Browser/Handlers/DownloadHandler.cs#L53). In `OnBeforeDownload`, I am comparing whether the `downloadItem.MimeType` of a to be downloaded item is equal to `MediaTypeNames.Application.Pdf`. If so, I cancel the download and open a new browser window, setting the download URL of the PDF as start URL.

This however obviously does not work (i.e. does not render the PDF in the newly opened window), but rather causes another download attempt which again will be canceled and so on. Thus the question, given that I have the URL to a PDF resource, how do I force the rendering of that document with the integrated PDFium reader, even if the server instructs the browser to download the PDF?

**Comment by aimcosoftware** [comment]
@dbuechel You can achieve it by using PDFJS, there is an example here: https://github.com/cefsharp/CefSharp/wiki/Embedding-PDF.JS-in-CEFSharp
As you're already using `OnBeforeDownload`, you are half way to using it, and you have complete control over the PDF rendering.

**Reply by dbuechel** [reply]
Thank you very much for the information @aimcosoftware. I have come across that particular wiki page while implementing, but would like to first try to implement the forced rendering of PDFs if there's any way to do so. Being able to do so would in my opinion have a few benefits (namely directly loading the document and thereby using the already available functionality in the browser engine / PDFium, thus keeping the PDF in memory [I presume] and not having to cache it locally, load it from there, cleanup afterwards etc.) when compared to having to completely manually handle PDFs, be it with pdfjs or PDFium.

**Reply by amaitland** [reply]
> Thus the question, given that I have the URL to a PDF resource, how do I force the rendering of that document with the integrated PDFium reader, even if the server instructs the browser to download the PDF?

There is no built in option for this that I'm aware of. Your welcome to ask on https://www.magpcss.org/ceforum/index.php for a second opinion.

The ideal solution would be modify the response headers, there's still an open upstream issue to add support for this (https://github.com/chromiumembedded/cef/issues/2785).

In theory you can disable the CEF request interceptor and switch to using the CDP for network interception, you could then use https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-continueResponse

As it stands currently downloading the file (to disk or memory depending on your requirements) then opening the `PDF` directly in the browser is likely the way forward. You obviously don't need to switch to pdf.js to achieve this.

**Reply by amaitland** [reply]
> The ideal solution would be modify the response headers, there's still an open upstream issue to add support for this ([chromiumembedded/cef#2785](https://github.com/chromiumembedded/cef/issues/2785)).

If you are going to spawn a new browser then you could possibly use a `RequestHandler`  and a `UrlRequestClient`. Where the `UrlRequestClient` effectively proxies the request, then modify the response headers. There would be overhead, so you'd only want to do this for requests that you absolutely had to.

**Reply by dbuechel** [reply]
> There is no built in option for this that I'm aware of.

That I was afraid of and coincides with my experience when trying to implement the functionality. Would be a nice to have feature, but as you stated that'd be something for the upstream CEF or Chromium projects I reckon.

> As it stands currently downloading the file (to disk or memory depending on your requirements) then opening the PDF directly in the browser is likely the way forward.

I'll thus try to download the file to memory and render it from there, as I really would like to be able to avoid having to cache it on disk and everything that comes with it. So given that I am already able to detect and intercept a PDF download, what would need to be done to render a PDF from memory? Do I need to implement and supply a custom resource handler (specifically here: https://github.com/SafeExamBrowser/seb-win-refactoring/blob/master/SafeExamBrowser.Browser/Handlers/ResourceHandler.cs#L71), or is there already an implementation available that I could use?

And thank you very much for the support, it is as always very much appreciated!

**Reply by dbuechel** [reply]
Any help would be greatly appreciated, I'm afraid I am stuck with having the PDF in memory and not knowing how to make the browser engine render it...

**Reply by amaitland** [reply, marked as answer]
If your going to keep the stream in memory then [Request Interception](https://github.com/cefsharp/CefSharp/wiki/General-Usage#request-interception) is one option.

There's a simplified option available using `RegisterResourceHandler`. Basically you tell the browser that for the given url it should load the stream. Then you load the url.

https://github.com/cefsharp/CefSharp/blob/2cb81aa943ed3a9787709bfa1764e91abf2bfbe3/CefSharp/WebBrowserExtensions.cs#L1003

**Reply by dbuechel** [reply]
Excellent, thanks a lot for the input. Yes, keeping it in memory would be preferable. I shall thus try to build a solution using the method mentioned, thanks again.
